{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ajk_utils.parse_fasta import parse_fasta\n",
    "from ajk_utils.orf_finder import get_orfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def parse_fasta(fasta_file):\n",
    "#     sequence = ''\n",
    "#     with open(fasta_file) as f:\n",
    "#         for line in f:\n",
    "#             if not line.startswith(\">\"):\n",
    "#                 sequence += line.strip()\n",
    "                \n",
    "#     return sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def get_genes(seq):\n",
    "#     first_start = sequence.index(\"ATG\") # stores position of first START codon\n",
    "#     stop_codons = ['TAG', 'TAA', 'TGA']\n",
    "#     start_positions = []\n",
    "#     for i in range(first_start, len(seq), 3):\n",
    "#         codon = seq[i:i+3]\n",
    "#         if codon == 'ATG':\n",
    "#             start_positions.append(i)\n",
    "#             #print(\"START\", codon, \"index =\", i)\n",
    "#         #if codon in stop_codons:\n",
    "#         #    print(\"STOP\", codon, \"index =\", i)\n",
    "                \n",
    "#     return start_positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# sequence = parse_fasta(\"sars_cov2.fasta\")\n",
    "# get_genes(sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genes(seq, min_bp, nested):\n",
    "    # Find forward strand orfs\n",
    "    sequence = parse_fasta(seq)\n",
    "    orfs_nested = get_orfs(sequence, min_bp, nested)\n",
    "    \n",
    "    # Create Reverse Complement strand\n",
    "    comp_base = {'A': 'T', 'T': 'A', 'G': 'C', 'C':'G'}\n",
    "    seq_list = list(sequence)\n",
    "    seq_list.reverse()\n",
    "    reverse = ''.join(seq_list)\n",
    "    r_comp = ''.join([comp_base[i] for i in reverse])\n",
    "    \n",
    "    # Find reverse strand orfs\n",
    "    orfs = get_orfs(r_comp, min_bp, nested)\n",
    "    comp_orfs_nested = [((len(r_comp) + 1 - i[1]), (len(r_comp) + 1 - i[0]), '-') \n",
    "                for i in orfs]\n",
    "    comp_orfs_nested.sort(key=lambda gene: (gene[0], gene[1], gene[2]))\n",
    "    \n",
    "    #Merge and sort both strands\n",
    "    gene_list = [(i[0], i[1], '+') for i in orfs_nested]\n",
    "    gene_list.extend(comp_orfs_nested)\n",
    "    gene_list.sort(key=lambda gene: (gene[0], gene[1], gene[2]))\n",
    "    \n",
    "    return gene_list\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene 1: (13, 108, '-')\n",
      "Gene 2: (266, 13480, '+')\n",
      "Gene 3: (425, 667, '-')\n",
      "Gene 4: (519, 692, '-')\n",
      "Gene 5: (594, 683, '+')\n",
      "Gene 6: (726, 878, '+')\n",
      "Gene 7: (966, 1094, '+')\n",
      "Gene 8: (1012, 1104, '-')\n",
      "Gene 9: (1394, 1570, '-')\n",
      "Gene 10: (1531, 1689, '-')\n",
      "Gene 11: (1605, 1736, '+')\n",
      "Gene 12: (1866, 1964, '+')\n",
      "Gene 13: (2293, 2457, '-')\n",
      "Gene 14: (2845, 2949, '-')\n",
      "Gene 15: (2958, 3203, '+')\n",
      "Gene 16: (3207, 3299, '+')\n",
      "Gene 17: (3362, 3466, '-')\n",
      "Gene 18: (3612, 3749, '+')\n",
      "Gene 19: (3948, 4100, '+')\n",
      "Gene 20: (4327, 4458, '-')\n",
      "Gene 21: (4707, 4832, '+')\n",
      "Gene 22: (4876, 5022, '-')\n",
      "Gene 23: (5103, 5213, '+')\n",
      "Gene 24: (5476, 5577, '-')\n",
      "Gene 25: (5690, 5779, '-')\n",
      "Gene 26: (5919, 6086, '+')\n",
      "Gene 27: (6156, 6347, '+')\n",
      "Gene 28: (6190, 6489, '-')\n",
      "Gene 29: (7235, 7336, '-')\n",
      "Gene 30: (7324, 7419, '-')\n",
      "Gene 31: (7548, 7706, '+')\n",
      "Gene 32: (7691, 7825, '-')\n",
      "Gene 33: (8001, 8099, '+')\n",
      "Gene 34: (8240, 8359, '-')\n",
      "Gene 35: (8800, 8949, '-')\n",
      "Gene 36: (9099, 9191, '+')\n",
      "Gene 37: (9398, 9493, '-')\n",
      "Gene 38: (9981, 10112, '+')\n",
      "Gene 39: (10043, 10177, '-')\n",
      "Gene 40: (10215, 10397, '+')\n",
      "Gene 41: (10276, 10371, '-')\n",
      "Gene 42: (10758, 10856, '-')\n",
      "Gene 43: (10951, 11085, '+')\n",
      "Gene 44: (11935, 12045, '-')\n",
      "Gene 45: (12157, 12255, '-')\n",
      "Gene 46: (12286, 12486, '-')\n",
      "Gene 47: (13118, 13273, '-')\n",
      "Gene 48: (13311, 13463, '+')\n",
      "Gene 49: (13362, 13550, '-')\n",
      "Gene 50: (13768, 21552, '+')\n",
      "Gene 51: (13908, 14027, '-')\n",
      "Gene 52: (13913, 14059, '-')\n",
      "Gene 53: (14651, 14749, '+')\n",
      "Gene 54: (14995, 15156, '-')\n",
      "Gene 55: (15461, 15664, '+')\n",
      "Gene 56: (15569, 15682, '-')\n",
      "Gene 57: (15793, 15888, '-')\n",
      "Gene 58: (16166, 16276, '+')\n",
      "Gene 59: (16203, 16295, '-')\n",
      "Gene 60: (16211, 16306, '-')\n",
      "Gene 61: (16493, 16600, '+')\n",
      "Gene 62: (16682, 16795, '+')\n",
      "Gene 63: (16802, 16939, '-')\n",
      "Gene 64: (16973, 17119, '+')\n",
      "Gene 65: (17014, 17106, '-')\n",
      "Gene 66: (17448, 17543, '-')\n",
      "Gene 67: (17509, 17628, '-')\n",
      "Gene 68: (17606, 17713, '+')\n",
      "Gene 69: (17982, 18185, '-')\n",
      "Gene 70: (18154, 18324, '-')\n",
      "Gene 71: (18410, 18520, '+')\n",
      "Gene 72: (19148, 19300, '+')\n",
      "Gene 73: (19716, 19856, '-')\n",
      "Gene 74: (19768, 19905, '-')\n",
      "Gene 75: (20168, 20260, '+')\n",
      "Gene 76: (20847, 21038, '-')\n",
      "Gene 77: (21536, 25381, '+')\n",
      "Gene 78: (21744, 21860, '+')\n",
      "Gene 79: (21936, 22196, '+')\n",
      "Gene 80: (22614, 22721, '+')\n",
      "Gene 81: (23077, 23322, '-')\n",
      "Gene 82: (23201, 23437, '-')\n",
      "Gene 83: (23403, 23510, '+')\n",
      "Gene 84: (23526, 23621, '+')\n",
      "Gene 85: (23683, 23787, '-')\n",
      "Gene 86: (23929, 24111, '-')\n",
      "Gene 87: (24051, 24167, '+')\n",
      "Gene 88: (24762, 24872, '+')\n",
      "Gene 89: (25102, 25242, '-')\n",
      "Gene 90: (25195, 25284, '+')\n",
      "Gene 91: (25332, 25445, '+')\n",
      "Gene 92: (25393, 26217, '+')\n",
      "Gene 93: (25457, 25579, '+')\n",
      "Gene 94: (25524, 25694, '+')\n",
      "Gene 95: (25589, 25807, '-')\n",
      "Gene 96: (25968, 26066, '+')\n",
      "Gene 97: (26072, 26167, '+')\n",
      "Gene 98: (26183, 26278, '+')\n",
      "Gene 99: (26184, 26342, '-')\n",
      "Gene 100: (26245, 26469, '+')\n",
      "Gene 101: (26523, 27188, '+')\n",
      "Gene 102: (26693, 26869, '+')\n",
      "Gene 103: (26807, 27004, '-')\n",
      "Gene 104: (27037, 27210, '-')\n",
      "Gene 105: (27202, 27384, '+')\n",
      "Gene 106: (27394, 27756, '+')\n",
      "Gene 107: (27643, 27771, '-')\n",
      "Gene 108: (27756, 27884, '+')\n",
      "Gene 109: (27894, 28256, '+')\n",
      "Gene 110: (28081, 28188, '+')\n",
      "Gene 111: (28274, 29530, '+')\n",
      "Gene 112: (28284, 28574, '+')\n",
      "Gene 113: (28731, 28835, '-')\n",
      "Gene 114: (28734, 28952, '+')\n",
      "Gene 115: (29210, 29341, '-')\n",
      "Gene 116: (29413, 29505, '-')\n",
      "Gene 117: (29558, 29671, '+')\n",
      "Gene 118: (29688, 29840, '-')\n"
     ]
    }
   ],
   "source": [
    "sars2_genes = get_genes('sars_cov2.fasta', 90, True)\n",
    "\n",
    "print('\\n'.join([f'Gene {i + 1}: {sars2_genes[i]}' \n",
    "               for i in range(len(sars2_genes))]))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
